text/0000-standard-lazy-types.md

```diff
@@ -0,0 +1,511 @@
+
+## Why not `Lazy` as a primitive?
+
+On the first look, it may seem like we don't need `OnceCell`, and should only provide `Lazy`.
+The critical drawback of `Lazy` is that it's not always possible to provide the closure at creation time.
+
+This is important for lazy fields:
+
+```rust
+struct Ctx {
+    config_path: PathBuf,
+    config: Lazy<String, ???>,
+}
+
+impl Ctx {
+    pub fn new(config_path: PathBuf) -> Ctx {
+        Ctx {
+            config_path,
+            config: Lazy::new(|| {
+                // We would like to write something like
+                // `fs::read_to_string(&self.config_path)`
+                // here, but we can't have access to `self`
+                ???
+            })
+        }
+    }
+}
+```
+
+Or for singletons, initialized with parameters:
+
+```rust
+use std::{env, io, sync::OnceCell};
+
+#[derive(Debug)]
+pub struct Logger { ... }
+
+static INSTANCE: OnceCell<Logger> = OnceCell::new();
+impl Logger {
+    pub fn global() -> &'static Logger {
+        INSTANCE.get().expect("logger is not initialized")
+    }
+    fn from_cli(args: env::Args) -> Result<Logger, std::io::Error> { ... }
+}
+
+fn main() {
+    let logger = Logger::from_cli(env::args()).unwrap();
+
+    // Note how we use locally-created value for initialization.
+    INSTANCE.set(logger).unwrap();
+
+    // use `Logger::global()` from now on
+}
+```
+
+## Why `OnceCell` as a primitive?
+
+It is possible to imagine a type, slightly more general than `OnceCell`:
+
+```rust
+struct OnceFlipCell<U, V> { ... }
+
+impl<U, V> OnceFlipCell<U, V> {
+    const fn new(initial_value: U) -> OnceFlipCell<U, V>;
+
+    fn get_or_init<F: FnOnce(U) -> V>(&self, f: F) -> &V;
+}
+
+type OnceCell<T> = OnceFlipCell<(), T>;
+```
+
+That is, we can store some initial state in the cell and consume it during initialization.
+In practice, such flexibility seems to be rarely required.
+Even if we add a type, similar to `OnceFlipCell`, having a dedicated `OnceCell` (which *could* be implemented on top of `OnceFlipCell`) type simplifies a common use-case.
+
+## Variations of `set`
+
+The RFC proposes "obvious" signature for the `set` method:
+
+```rust
+fn set(&self, value: T) -> Result<(), T>;
+```
+
+Note, however, that `set` establishes an invariant that the cell is initialized, so a more precise signature would be
+
+```rust
+fn set(&self, value: T) -> (&T, Option<T>);
+```
+
+To be able to return a reference, `set` might need to block a thread.
+For example, if two threads call `set` concurrently, one of them needs to block while the other moves the value into the cell.
+It is possible to provide a non-blocking alternative to `set`:
+
+```rust
+fn try_set(&self, value: T) -> Result<&T, (Option<&T>, T)>
+```
+
+That is, if value is set successfully, a reference is returned.
+Otherwise, ther the cell is either fully initialized, and a reference is returned as well, or the cell is being initialized, and no valid reference exist yet.
```

Comment:
```suggestion
Otherwise, the cell is either fully initialized, and a reference is returned as well, or the cell is being initialized, and no valid reference exist yet.
```